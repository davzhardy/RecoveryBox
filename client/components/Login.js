import React, { useEffect, useState } from 'react';
import { StyleSheet, View, TouchableOpacity, Text } from 'react-native';
import { BoldAppText } from '../styles/text'
import {useDispatch, useSelector, connect } from "react-redux";
import ApiService from '../ApiService'
import colors from '../styles/colors'
import { StackActions } from '@react-navigation/native';

import * as Google from 'expo-google-app-auth';
import { androidClientId, iosClientId } from '../config/secret.js'
  
function LoginScreen ({ navigation }) {

  const dispatch = useDispatch();
  const [warning, setWarning] = useState(false);
  const [jwt, setJwt] = useState('');
  const [googleOauth, setGoogleOauth] = useState('');

  const oAuthSignIn = async () => {
    try {
      //TODO: check if this also work for android phones
      const result = await Google.logInAsync({
        androidClientId: androidClientId,
        iosClientId: iosClientId,
        scopes: ['profile', 'email'],
        redirectUrl: `com.mazethernandez.recoverybox:/oauth2redirect/google` 
      });
      if (result.type === 'success') {
        console.log('user', result);
        await setGoogleOauth(result.user.id);
        receiveJwt({idToken: result.idToken});
      } else {
      console.log('TYPE', result.type);
      console.log('cancelled');
      }
    } catch (e) {
    console.log('error', e);
    }
  }
  
  function receiveJwt (userToken) {
    ApiService.getJwt(userToken)
      .then(serverResponse => {
        setJwt(serverResponse.accessToken)
        //FIXME: this is temporary solution to ensure jwt stored globally but should be integrated with redux map
        dispatch({
          type: 'SAVE_JWT',
          payload: serverResponse.accessToken
        })
        // )
      })
  }

  useEffect(() => {
    if (jwt.length && googleOauth.length) {
      receiveInfoandData(googleOauth, jwt)
      navigation.dispatch(
        StackActions.replace('Home')
      )
    }
  }, [jwt])

  function receiveInfoandData (userId, accessToken) {
    ApiService.getUserInfo(userId, accessToken)
    .then(data => {
      let dispatchtoUser = {
        id: data[0].id,
        email: data[0].email,
        firstName: data[0].firstName,
        lastName: data[0].lastName,
      }
      if(data[0].Data.length) {
        for (let i of data[0].Data) {
          i.date = Number(i.date)
          let parseMoods = i.moods.replace(/[\[\]',"]+/g,'')
          let arrayMoods;
          if (parseMoods.length) arrayMoods = parseMoods.split(' ')
          if (arrayMoods===undefined) arrayMoods = []
          i.moods = arrayMoods
          i.suggestions = eval(i.suggestions)
        }
      }
      let dispatchtoHistoricalData = data[0].Data
      console.log('dispatchtoHistoricalData:', dispatchtoHistoricalData)
      dispatch({
        type: 'UPDATE_USERINFO',
        payload: dispatchtoUser
      })
      dispatch({
        type: 'CREATE_HISTORICALDATA',
        payload: dispatchtoHistoricalData
      })
    })
  }

  return (
    <View style={styles.container}>
      <BoldAppText style={styles.logo}>RecoveryBox</BoldAppText>
      <TouchableOpacity style={styles.oAuthButton} onPress={() => oAuthSignIn()}>
        <Text style={styles.text}>LOGIN WITH GOOGLE</Text>
      </TouchableOpacity>
      <View>
        { 
          warning ? <Text>Please submit both a username and password</Text> : null
        }
      </View>
    </View>
  );
}
  
const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.platinum,
  },
  inputView:{
    width:"80%",
    backgroundColor: colors.blue,
    borderRadius:25,
    height:50,
    marginBottom:20,
    justifyContent:"center",
    padding:20,
  },
  logo: {
    color: colors.orange, 
    fontSize: 40,
    marginBottom: 60,
  },
  inputText: {
    height:50,
    color:'white',
    textAlign: 'center',
  },
  button: {
    width:"60%",
    backgroundColor: colors.orange,
    borderRadius:25,
    height:50,
    alignItems:"center",
    justifyContent:"center",
    marginTop:30,
    marginBottom:10,
  },
  oAuthButton: {
    width:"60%",
    backgroundColor: colors.orange,
    borderRadius:25,
    height:50,
    alignItems:"center",
    justifyContent:"center",
    marginTop:30,
    marginBottom:10,
  },
  text: {
    fontFamily: 'Montserrat_500Medium',
    fontStyle: "normal",
    fontSize: 16,
    color: "white",
  }
});

export default LoginScreen;