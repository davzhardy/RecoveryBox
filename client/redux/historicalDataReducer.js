const initialState = {

  1600300800000: {
    date: 1600300800000,
    meetings: 2,
    feeling: 5,
    moods: ['tired', 'happy', 'bored'],
    suggestions: ['Meetings', 'Prayer', 'Meditation', 'Useful', 'Kind to Myself', 'Gratitude', 'Step 10s', 'Inventory', 'Called a Newcomer'],
  },
  1600387200000: {
    date: 1600387200000,
    meetings: 1,
    feeling: 4,
    moods: ['tired', 'happy', 'down'],
    suggestions: ['Gratitude', 'Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  },
  1600473600000: {
    date: 1600473600000,
    meetings: 0,
    feeling: 3,
    moods: ['elated', 'happy', 'sad'],
    suggestions: ['Inventory', 'Called a Newcomer', 'Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  },
  1600560000000: {
    date: 1600560000000,
    meetings: 2,
    feeling: 2,
    moods: ['joyful', 'happy', 'sad'],
    suggestions: ['Kind to Myself']
  },
  1600646400000: {
    date: 1600646400000,
    meetings: 5,
    feeling: 5,
    moods: ['bored', 'happy', 'sad'],
    suggestions: ['Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  },
  1600819200000: {
    date: 1600819200000,
    meetings: 2,
    feeling: 10,
    moods: [],
    suggestions: []
  },
  1600905600000: {
    date: 1600905600000,
    meetings: 0,
    feeling: 9,
    moods: ['stressed', 'happy', 'sad'],
    suggestions: ['Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  },
  1601078400000: {
    date: 1601078400000,
    meetings: 5,
    feeling: 7,
    moods: ['tired', 'happy', 'sad'],
    suggestions: ['Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  }
}

// TODO refactor this to some kind of array so the data is more easily accessible

const historicalDataReducer = (state = initialState, action) => {
  switch (action.type) {
    case "UPDATE_HISTORICALDATA_WITH_DAILYINFO":
      return {...state, key: action.payload}
    case "UPDATE_HISTORICALDATA_WITH_HISTORICALINFO":
      return {...state, newkey: action.payload}
    case "CREATE_HISTORICALDATA":
      return {...state, ...action.payload}
    default:
      return state;
  }
}

export default historicalDataReducer;


// TODO delete this historical data when im 100% sure I dont need it 
 // 1600300800000: {
  //   date: 1600300800000,
  //   meetings: 2,
  //   feeling: 5,
  //   moods: ['tired', 'happy', 'bored'],
  //   suggestions: ['Meetings', 'Prayer', 'Meditation', 'Useful', 'Kind to Myself', 'Gratitude', 'Step 10s', 'Inventory', 'Called a Newcomer'],
  // },
  // 1600387200000: {
  //   date: 1600387200000,
  //   meetings: 1,
  //   feeling: 4,
  //   moods: ['tired', 'happy', 'down'],
  //   suggestions: ['Gratitude', 'Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  // },
  // 1600473600000: {
  //   date: 1600473600000,
  //   meetings: 0,
  //   feeling: 3,
  //   moods: ['elated', 'happy', 'sad'],
  //   suggestions: ['Inventory', 'Called a Newcomer', 'Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  // },
  // 1600560000000: {
  //   date: 1600560000000,
  //   meetings: 2,
  //   feeling: 2,
  //   moods: ['joyful', 'happy', 'sad'],
  //   suggestions: ['Kind to Myself']
  // },
  // 1600646400000: {
  //   date: 1600646400000,
  //   meetings: 5,
  //   feeling: 5,
  //   moods: ['bored', 'happy', 'sad'],
  //   suggestions: ['Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  // },
  // 1600819200000: {
  //   date: 1600819200000,
  //   meetings: 2,
  //   feeling: 10,
  //   moods: [],
  //   suggestions: []
  // },
  // 1600905600000: {
  //   date: 1600905600000,
  //   meetings: 0,
  //   feeling: 9,
  //   moods: ['stressed', 'happy', 'sad'],
  //   suggestions: ['Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  // },
  // 1601078400000: {
  //   date: 1601078400000,
  //   meetings: 5,
  //   feeling: 7,
  //   moods: ['tired', 'happy', 'sad'],
  //   suggestions: ['Prayer', 'Meditation', 'Useful', 'Kind to Myself']
  // },