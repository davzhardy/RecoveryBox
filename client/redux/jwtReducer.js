const initialState = {
	jwt: '' ,
}

const jwtReducer = (state = initialState, action) => {
	switch (action.type) {
		case 'SAVE_JWT': 
		return {
			...state,
			jwt: action.payload
		}
		case 'DELETE_JWT': 
		return {
			...state,
			jwt: '',
		}
		default:
      return state;
	}
}

export default jwtReducer;