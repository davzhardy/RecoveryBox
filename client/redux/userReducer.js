const initialState = {
  id: false,
  email: false,
  username: false,
  password: false,
  firstName: false,
  lastName: false,
  registrationDate: false,
}

const userReducer = (state = initialState, action) => {
  switch (action.type) {
    case "UPDATE_USERNAME":
      return {
        ...state,
        username: action.payload
      }
    case "UPDATE_USERINFO":
      return {
        ...state,
        id: action.payload.id,
        email: action.payload.email,
        username: action.payload.username,
        firstName: action.payload.firstName,
        lastName: action.payload.lastName,
        registrationDate: action.payload.registrationDate,
      }
    default:
      return state;
  }
}

export default userReducer;