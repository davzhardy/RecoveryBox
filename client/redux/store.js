import { createStore } from "redux";
import { combineReducers } from "redux";
import dailyInfoReducer from './dailyInfoReducer'
import settingsReducer from './settingsReducer'
import userReducer from './userReducer'
import quoteReducer from './quoteReducer'
import historicalDataReducer from './historicalDataReducer'
import helperReducer from './helperReducer'
import calendarHistoricalDayReducer from './calendarHistoricalDayReducer'
import jwtReducer from './jwtReducer';

const rootReducer = (combineReducers)({
  user: userReducer,
  dailyInfo: dailyInfoReducer,
  settings: settingsReducer,
  dailyQuote: quoteReducer,
  historicalData: historicalDataReducer,
  helper: helperReducer,
  calendarHistoricalDay: calendarHistoricalDayReducer,
  jwt: jwtReducer,
})

const store = createStore(rootReducer);

export default store
