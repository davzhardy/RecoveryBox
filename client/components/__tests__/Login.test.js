import React from 'react';
import { Provider } from "react-redux";
import { createStore } from "redux";
import userReducer from "../../redux/userReducer";
import renderer from 'react-test-renderer';
import Login from '../Login';

const initialState = {
  user: {
    id: false,
    email: false,
    username: false,
    password: 'TODO: replace me',
    firstName: false,
    lastName: false,
    registrationDate: false,
  }
}

const store = createStore(userReducer, initialState);

test('renders correctly', () => {
  const tree = renderer.create(<Provider store = {store}><Login/></Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});