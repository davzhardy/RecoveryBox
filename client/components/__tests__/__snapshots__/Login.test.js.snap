// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#DFE2E2",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FF8787",
        "fontFamily": "Montserrat_700Bold",
        "fontSize": 40,
        "fontStyle": "normal",
        "marginBottom": 60,
      }
    }
  >
    RecoveryBox
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#2F80ED",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "marginBottom": 20,
        "padding": 20,
        "width": "80%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Enter a username"
      placeholderTextColor="#DFE2E2"
      rejectResponderTermination={true}
      style={
        Object {
          "color": "white",
          "height": 50,
          "textAlign": "center",
        }
      }
      textContentType="username"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#2F80ED",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "marginBottom": 20,
        "padding": 20,
        "width": "80%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Enter a password"
      placeholderTextColor="#DFE2E2"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "color": "white",
          "height": 50,
          "textAlign": "center",
        }
      }
      underlineColorAndroid="transparent"
      value="TODO: replace me"
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FF8787",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "marginBottom": 10,
        "marginTop": 30,
        "opacity": 1,
        "width": "60%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "Montserrat_500Medium",
          "fontSize": 16,
          "fontStyle": "normal",
        }
      }
    >
      LOGIN
    </Text>
  </View>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FF8787",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "marginBottom": 10,
        "marginTop": 7,
        "opacity": 1,
        "width": "60%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "Montserrat_500Medium",
          "fontSize": 16,
          "fontStyle": "normal",
        }
      }
    >
      REGISTER
    </Text>
  </View>
  <View />
</View>
`;
